File: 2020/day_02/day_02.py
def clean_data(line):
    parts = line.split()
    p1 = parts[0]
    minimum, maximum = map(int, p1.split('-'))
    return minimum, maximum, parts[1][0], parts[2]


with open('in') as f:
    data = list(map(clean_data, f.readlines()))


def part_one():
    nr_valid = 0
    for (m, M, c, p) in data:
        count = p.count(c)
        if count >= m and count <= M:
            nr_valid += 1

    return nr_valid


def part_two():
    nr_valid = 0
    for (i, j, c, p) in data:
        count = (p[i - 1] == c) + (p[j - 1] == c)
        if count == 1:
            nr_valid += 1

    return nr_valid


if __name__ == '__main__':
    print(f'Part One: {part_one()}')
    print(f'Part Two: {part_two()}')

File: 2020/day_02/day_02.rs
use std::fs;

fn main() {
    let f = fs::read_to_string("in").unwrap();
    let data: Vec<_> = f.lines().collect();

    println!("Part One: {}", part_one(&data));
    println!("Part Two: {}", part_two(&data))
}

fn clean_line<'a>(line: &'a str) -> (i32, i32, char, &'a str) {
    let parts: Vec<_> = line.split_whitespace().collect();
        
    let part1: Vec<i32> = parts[0].split("-").map(|x| x.parse().unwrap()).collect();
    let min = part1[0];
    let max = part1[1];

    let letter = parts[1].chars().next().unwrap();
    let pass = parts[2];

    (min, max, letter, pass)
}

fn part_one(data: &Vec<&str>) -> i32 {
    let mut valid = 0;
    
    for line in data.iter().map(|x| x.trim()) {
        let (min, max, letter, pass) = clean_line(line);

        let c: i32 = pass.chars().filter(|x| *x == letter).count() as i32;
        if c >= min && c <= max {
            valid += 1;
        }
    }

    valid
}

fn part_two(data: &Vec<&str>) -> i32 {
    let mut valid = 0;
    
    for line in data.iter().map(|x| x.trim()) {
        let (min, max, letter, pass) = clean_line(line);

        let contains1 = pass.chars().nth((min-1) as usize).unwrap() == letter;
        let contains2 = pass.chars().nth((max-1) as usize).unwrap() == letter;

        if (contains1 && !contains2) || (!contains1 && contains2) {
            valid += 1;
        }
    }

    valid
}
